Updated ProxyFactory to generate proxies compatible with PHP 7 TypeHint

vermi0ffh opened the pull request:
Actual ProxyFactory don't works well with PHP7 type hint et return type.

Examples :

`public function getId(): ?string
{
  return $this->id;
}`

Old generator gives :
`public function getId()
{
  $this->__load();
  return parent::getId();
}`

This lead to a compilation error because getId() is not compatible with he original getId() (no return type).

With this patch, the result is :
`public function getId(): ?string
{
  $this->__load();
  return parent::getId();
}`

No more compilation error. It's backward compatible with the existing code and PHP < 7 (not tested).

Commit:
Updated ProxyFactory to generate proxies compatible with PHP 7 TypeHint (prevent some compilation errors)

Commit:
Patch to make the code retro-compatible with PHP < 7

Ocramius commented on lib/Doctrine/ODM/CouchDB/Proxy/ProxyFactory.php:
```diff
@@ -187,12 +188,14 @@ private function generateMethods(ClassMetadata $class)
-                        $parameterString .= 'array ';
-                    }
+	                // We need to pick the type hint class too
+	                if (($paramClass = $param->getClass()) !== null) {
```
This code should rather pint at the `doctrine/common` proxy-factory, or use the ProxyManager component, like is done in https://github.com/doctrine/orm/pull/6719

vermi0ffh replied:
I pushed a new commit to use the new ProxyManager component. Did some tests to help travis run go smoothly, but it's not very efficient. Any comment on the new code ?
I did not started a full code review/refactoring as it's working, but I saw a lot a problems that need to be addressed.

Commit:
- Updated ProxyFactory to use ProxyManager
- Modification of Unit Tests
- Updated requirements and dependancies
- Added my name to the author list
- Warning : transcient fields code should be broken
- All unit tests pass with CouchDB 1.7

Commit:
- PHP 7.2 as minimum requirement (needed by ProxyManager)

Commit:
- Updated travis configuration, now using maintened PHP versions

Commit:
- PHP 7.4 is a bit too new for now...

Commit:
- Git got all messed up

Commit:
- Typo in file name

Commit:
- Removing unit test with proxy class name (not predicable enough)

vermi0ffh commented:
Any plan to merge this pull request one day ?
